lib/vanagon/project.rb

```diff
@@ -82,6 +82,32 @@ def get_files
       files.flatten.uniq
     end
 
+    # Returns a filtered out set of components only including those
+    # components necessary to build a specific component. This is a
+    # recursive function that will call itself until it gets to a
+    # component with no build requirements
+    #
+    # @param name [String] name of component to add. must be present in configdir/components and named $name.rb currently
+    def filter_component(name)
+      components_to_include = []
+      filter_component = get_component(name)
+      return nil if filter_component.nil?
+      components_to_include << filter_component
+      filter_component.build_requires.each do |build_requirement|
+        unless get_component(build_requirement).nil?
+          components_to_include += filter_component(build_requirement)
+        end
+      end unless filter_component.build_requires.nil?
+      components_to_include
+    end
+
+    def get_component(name)
+      unless @components.select { |comp| comp.name.to_s == name.to_s }.nil?
+        return @components.select { |comp| comp.name.to_s == name.to_s }.pop
+      end
+      nil
+    end
```

Comment:
Cross-checking confirmed that `@project.components` is always an Array. So `#select` is always safe, and yes, it is the way to pluck a single item out of an array. But `#select` is going to result in an empty array if your item isn't in the array and it will always return an Array.

Here's your method as written:

```ruby
def get_component(name)
  unless @components.select { |comp| comp.name.to_s == name.to_s }.nil?
    return @components.select { |comp| comp.name.to_s == name.to_s }.pop
  end
  nil
end
```

`@components.select { |comp| comp.name.to_s == name.to_s }` is **always** an Array. So testing `#nil?`will always fail because an Array is never `nil`. If you wanted to test the value of the first (only?) component, you'd use `.first`:

```ruby
def get_component(name)
  @components.select { |comp| comp.name.to_s == name.to_s }.first
end
```

Would this do the thing you intended it to do? What about:

```ruby
def get_component(name)
  comps = @components.select { |comp| comp.name.to_s == name.to_s }
  raise "WHY ARE THERE SO MANY COMPONENTS WITH THE SAME NAME" if comps.size > 1
  comps.first
end